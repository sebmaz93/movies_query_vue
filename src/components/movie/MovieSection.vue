<script setup lang="ts">
// components
import MovieComponent from '@components/movie/Movie.vue'
import FiltersSection from '@components/_common/filter/FiltersSection.vue'

// hooks
import { useQuery } from '@hooks/useQuery';

// services
import { fetchMovies } from '@services/api/movie';

// types
import { MoviesRes, MovieReqParams } from '@services/api/movie/types';

const { data: movies, error, isLoading, fetch: getMovies } = useQuery<MoviesRes, MovieReqParams>('MOVIES_GET', fetchMovies);

</script>

<template>
    <section class="main-container">
        <FiltersSection :getMovies="getMovies" :hasNext="Boolean(movies?.has_next_page)" />
        <div class="loader" v-if="isLoading">Loading...</div>
        <h2 v-else-if="error">Error</h2>
        <div class="movies" v-if="!isLoading && movies?.results">
            <MovieComponent
                v-if="movies.results.length > 0"
                v-for="movie in movies.results"
                :movie="movie"
                :key="movie.id"
            />
            <div v-else>No Movies Found.</div>
        </div>
    </section>
</template>

<style lang="scss" scoped>
.main-container {
    height: 100%;
    width: 100%;
    display: flex;
    flex-direction: column;
    align-items: center;
    gap: 32px;
}

.loader {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    font-size: 32px;
}
.movies {
    display: flex;
    justify-content: center;
    flex-wrap: wrap;
    gap: 24px;
    height: 100%;
    width: 100%;
}
</style>
