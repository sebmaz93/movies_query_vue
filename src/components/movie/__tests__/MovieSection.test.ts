import { render, fireEvent } from '@testing-library/vue';
import MovieSection from '../MovieSection.vue';
import { nextTick } from 'vue';
import moviesApi from '@services/api/movie';

describe('Render MoviesSection', () => {
  beforeEach(() => jest.clearAllMocks());

  it('should not have movie in the list', async () => {
    const { queryByTestId } = render(MovieSection);
    const api = jest.spyOn(moviesApi, 'fetchMovies').mockImplementation(
      () =>
        new Promise((resolve) =>
          resolve({
            data: {
              page: 1,
              total_pages: 2,
              has_next_page: true,
              results: [
                {
                  id: 1,
                  title: 'mymovie',
                  img: 'poster',
                  genre: [
                    {
                      id: 2,
                      title: 'sci-fi',
                    },
                  ],
                  actors: [
                    {
                      id: 3,
                      name: 'seb',
                    },
                  ],
                  is_series: false,
                  release_date: 2021,
                },
              ],
              total_results: 14,
            },
          } as any)
        )
    );

    expect(queryByTestId('movies-container')).not.toBeInTheDocument();
    expect(api.mock.calls).toMatchSnapshot();
  });

  it('should show loader then have movies in list', async () => {
    const api = jest.spyOn(moviesApi, 'fetchMovies').mockImplementation(
      () =>
        new Promise((resolve) =>
          resolve({
            data: {
              page: 1,
              total_pages: 2,
              has_next_page: true,
              results: [
                {
                  id: 1,
                  title: 'mymovie',
                  img: 'poster',
                  genre: [
                    {
                      id: 2,
                      title: 'sci-fi',
                    },
                  ],
                  actors: [
                    {
                      id: 3,
                      name: 'seb',
                    },
                  ],
                  is_series: false,
                  release_date: 2021,
                },
              ],
              total_results: 14,
            },
          } as any)
        )
    );
    const { getByTestId } = render(MovieSection);

    fireEvent.click(getByTestId('search-btn'));
    await nextTick();

    expect(getByTestId('loader')).toBeInTheDocument();

    await new Promise((resolve) => setTimeout(resolve, 1500));
    await nextTick();

    expect(getByTestId('movies-container')).toBeInTheDocument();
    expect(getByTestId('movies-container').childElementCount).toBe(8);
    expect(api.mock.calls).toMatchSnapshot();
  });
});
