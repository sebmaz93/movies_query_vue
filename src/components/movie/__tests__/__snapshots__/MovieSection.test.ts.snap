// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render MoviesSection should not have movie in the list 1`] = `Array []`;

exports[`Render MoviesSection should show loader then have movies in list 1`] = `Array []`;
